invites: Allow users to invite without specifying any stream to join.

sahil839 opened the pull request:
We now allow users to invite without specifying any stream to join.
In such cases, the user would join the default streams, if any, during
the process of account creation after accepting the invite.

It is also fine if there are no default streams and user isn't subscribed to any stream initially.

<!-- Describe your pull request here.-->
 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![invite-without-streams](https://user-images.githubusercontent.com/35494118/236596699-18af109d-d3e1-4c2d-8dfd-204bf581051d.gif)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
I think if you uncheck all the boxes, the user should be invited with no subscribed streams. You made an explicit choice to *not* invite the user to default streams, so it doesn't make sense to still add the invited user to default streams at that point.

alya commented:
Chatting with @timabbott just now, we can have a checkbox under "Streams they should join":

[ ] Default streams for this organization

If it's checked, we hide the checkboxes part altogether, and use default streams at the time that the user joins. If it's unchecked, we show the checkboxes part and use the specific set of streams the user has specified.

sahil839 commented:
> I think if you uncheck all the boxes, the user should be invited with no subscribed streams. You made an explicit choice to _not_ invite the user to default streams, so it doesn't make sense to still add the invited user to default streams at that point.

This was the behavior previously for multi-use invites where we allowed sending without selecting any streams.

Adding a new checkbox to subscribe to default streams sounds good. Will update the PR.

sahil839 commented on zerver/actions/create_user.py:
```diff
@@ -182,9 +182,12 @@ def process_new_human_user(
+    if len(streams) == 0 and not user_was_invited:
         streams = get_default_subs(user_profile)
 
     for default_stream_group in default_stream_groups:
```
I am not sure what we want do for streams in `default_stream_groups`. We previously added the streams in `default_stream_groups` irrespective of the streams passed during invite.

sahil839 commented on web/templates/invite_user_modal.hbs:
```diff
@@ -56,6 +56,15 @@
         <label>{{t "Streams they should join" }}</label>
         <div id="streams_to_add">
+            {{#if show_select_default_streams_option}}
+            <div class="select_default_streams new-style">
```
Have used `new-style` class for now as it seemed easiest way to make the checkbox styling consistent with others. Would focus on removing the class, once I get done with release related work.

timabbott replied:
Sounds reasonable.

sahil839 commented:
Updated the PR -
- The user is now not subscribed to any stream, including default streams, if no stream was selected while inviting.
- Added a couple of commits to fix a typo and add test.
- Added checkbox to select default streams. Note that we do not show the option if there are no default streams in the realm.

sahil839 commented:
This might need some changes in design. Should we make "Streams they should join" look like a heading?
![Screenshot from 2023-05-09 16-57-55](https://github.com/zulip/zulip/assets/35494118/e0aaa166-d37c-4096-b168-fab2f66b63bd)
![invite-default-streams](https://github.com/zulip/zulip/assets/35494118/73e0065f-3eec-4c16-9641-7cecc6301011)

alya commented:
> This might need some changes in design. Should we make "Streams they should join" look like a heading?

I think it's OK, since it's consistent with the other labels.

alya commented:
If you change the default streams, it looks to me in manual testing like the default streams at the time that the invite was sent are used. My understanding of the plan for this was:

> I think we use the default streams as of the time the invite was accepted currently, and it's fine to leave it that way; I think that's what users will generally prefer in the case that the user doesn't super have control over it.

But I don't feel super strongly either way about this detail.

alya commented:
Looks good to me otherwise! I really like that we end up hiding (by default) the not-super-stylish checkboxes extravaganza as a result of this change. :)

alya commented:
Anyway, @timabbott I think you can take a look at this point.

timabbott commented:
Merged after a tiny edit to the API changelog, thanks @sahil839!

timabbott commented:
I think it would likely be an improvement to make this new checkbox do ""

In any case, @sahil839 you can now do a PR backporting this to 6.x.

sahil839 commented:
> I think it would likely be an improvement to make this new checkbox get the default streams at the time the user accepts the invite, since that better fits user expectations... but that might require some additional fiddling and I don't want to block on that.
> 
> In any case, @sahil839 you can now do a PR backporting this to 6.x.

I missed the point about subscribing to the default streams at the time user accepts the invite. For this might need to send a special value to API, probably in a separate argument, like` subscribe_to_default_streams_only` and add this new field to `PreregistrationUser` and `MultiuseInvite` objects as well.

Will open a PR for backporting soon.
